### ISSUE TITLE
HIGH: [Polish] [$500] Allow workspace admins to delete system messages

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#37119]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.43-14
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1708637188991179

## Action Performed:
1. Open a workspace chat /room which has system message and admin of the workspace
2. Try to delete the message as admin

## Expected Result:
Able to delete the message like any other message

## Actual Result:
Unable to delete the message

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ae45c744200c5267</li>
        <li>Upwork Job ID: 1760834635148304384</li>
        <li>Last Price Increase: 2024-03-22</li>
    </ul>
</details>

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Allow workspace admins to delete system messages

### What is the root cause of that problem?

We are currently not allowed to do it.

### What changes do you think we should make in order to solve the problem?
Is it duplicate solution? I believe it isn't, my solution consider this check:

```
        (reportAction?.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE && !isAdmin)
```
And add optional whisper message delete behavior.

Back end fix is required.

We can change this code block:

https://github.com/Expensify/App/blob/88a3ffe9440548d04c32bec2c889b8bb063a29b3/src/libs/ReportUtils.ts#L1274-L1285

To:

**If we want to allow the deletion of all system messages.**

```
    if (
        (reportAction?.actionName !== CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT && !isAdmin) ||
        reportAction?.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        (reportAction?.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE && !isAdmin)
    ) {
        return false;
    }
```

**If we only allowing only specific action types, we can define `isIncludedInAllowDeleteActionList`.**

```
    const isAdmin = policy?.role === CONST.POLICY.ROLE.ADMIN && !isEmptyObject(report) && !isDM(report);

    // Define the list of allowed delete actions
    const allowDeleteActionList: string[] = [
        CONST.REPORT.ACTIONS.TYPE.ACTIONABLEMENTIONWHISPER,
        CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT,
        // Add other action types as needed
    ];

    const isIncludedInAllowDeleteActionList = reportAction?.actionName && allowDeleteActionList.includes(reportAction?.actionName) || false;
    
    if (
        (!isIncludedInAllowDeleteActionList && !isAdmin) ||
        reportAction?.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        (reportAction?.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE && !isAdmin)
    ) {
        return false;
    }

    return (isActionOwner || isAdmin || isWhisperOnlyVisibleByUser) && isIncludedInAllowDeleteActionList;

```

**If we want to allow non-admin users to delete their own whisper messages**

```
    const isAdmin = policy?.role === CONST.POLICY.ROLE.ADMIN && !isEmptyObject(report) && !isDM(report);

    // Define the list of allowed delete actions
    const allowDeleteActionList: string[] = [
        CONST.REPORT.ACTIONS.TYPE.ACTIONABLEMENTIONWHISPER,
        CONST.REPORT.ACTIONS.TYPE.ADDCOMMENT,
        // Add other action types as needed
    ];

    const isIncludedInAllowDeleteActionList = reportAction?.actionName && allowDeleteActionList.includes(reportAction?.actionName) || false;
    
    if (
        (!isIncludedInAllowDeleteActionList && !isAdmin) ||
        reportAction?.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE ||
        ReportActionsUtils.isCreatedTaskReportAction(reportAction) ||
        (reportAction?.actorAccountID === CONST.ACCOUNT_ID.CONCIERGE && !isAdmin)
    ) {
        return false;
    }

    const whisperedToAccountIDs = reportAction?.whisperedToAccountIDs || [];
    const isWhisper = whisperedToAccountIDs.length > 0;
    const isWhisperOnlyVisibleByUser = isWhisper && isCurrentUserTheOnlyParticipant(whisperedToAccountIDs);

    return (isActionOwner || isAdmin || isWhisperOnlyVisibleByUser) && isIncludedInAllowDeleteActionList;

```
### What alternative solutions did you explore? (Optional)

**Comment by miljakljajic** [comment]
Are we sure they should be able to do this? @mountiny what do you think?

**Comment by mountiny** [comment]
@miljakljajic I think we might want to explore if admins should be able to delete _anything_, I dont think that is the case, but this issue was flagged by DB so maybe it should be brought up in the VSB channel

**Comment by miljakljajic** [comment]
Thank you for adding it to VSB!

**Comment by miljakljajic** [comment]
Waiting on an internal engineer to pick this up

**Comment by miljakljajic** [comment]
Still waiting

**Comment by miljakljajic** [comment]
Increased the status

**Comment by miljakljajic** [comment]
still waiting for a volunteer

**Comment by miljakljajic** [comment]
still waiting for an internal engineer - called out as a fire in #vsb: https://expensify.slack.com/archives/C066HJM2CAZ/p1714786700283159

**Comment by quinthar** [comment]
> do we want to allow admins to delete essentially anything? even system messages might be iffy - would that include like workspace logs?
Yes admins should be able to delete anything, but our delete is a "soft delete" -- it's still in the database, but it's just not shown.

**Comment by dangrous** [comment]
Ah okay - so this is a bit more complicated than I thought, since just allowing admins to delete system messages doesn't actually work properly - e.g. adding `deleted`, empty `html`, etc. etc. to a policy change log doesn't actually delete the log in the front end, since the action is built up from other properties.

Is there a list somewhere of all the types of system messages we want to be able to delete? It might take some refining on an individual basis to make sure each doesn't show on the frontend but is, as you said, still in the database.

**Comment by miljakljajic** [comment]
I'm not 100% clear on why we'd allow an admin to delete workspace logs? Doesn't that make the logs unreliable as a record for other admins? Is there some more context outside of this [post](https://expensify.slack.com/archives/C049HHMV9SM/p1708637188991179) that I can catch up on?

**Comment by quinthar** [comment]
@miljakljajic They are still there in the database, just not distracting the user.   I personally do this all the time for Slack, to clean up random system messages that are distracting from what I want to do.  The policy audit log is a _feature_, not something we are forcing on users.  If for some reason they want to delete it, I think that's fine.  It's their workspace, after all.  

THAT SAID, I do think we should create a feature that admins can use to basically "show all deleted comments", and that would reveal these.  In fact... I could see an argument that admins should see a WhatsApp style `<deleted message>` line there, so they are aware that something was deleted... But that feels out of scope for this.

@dangrous What about adding an optional `html` element that overrides the system message auto-generated HTML... and then we just set it to `""` to delete the message?

**Comment by dangrous** [comment]
Yeah I think we can make that work! I think if we do that the messages would still be in Onyx so technically accessible from the front end if you did digging, but I think that should be fine. I'll also have to make sure that the blank html gets sent from the backend, I'm not sure it would right now with `buildLastMessageItem` but will make sure. I can get going on that (and the front end bits) next week!

**Comment by dangrous** [comment]
Work in progress [here](https://github.com/Expensify/Auth/pull/11044). This allows the correct behavior on the backend, but bringing the changes out to the frontend is still forthcoming

**Comment by dangrous** [comment]
Okay, necessary PHP and Auth changes are up. Just need to show the option on the front end for admins, do a little more testing, and then we should be good! Ideally this will get done this week.

**Comment by dangrous** [comment]
One last question for the group cc @quinthar @miljakljajic - can I get a clarification on what system messages specifically we're trying to delete? I know it at least includes policy changelogs, but is there anything else?

Asking because my initial solutions allow for admins to delete all actions, and I'm wondering if we should restrict it at all. For example, they can delete IOUs. Is that fine? If not, I'd love a list of the various actions that count as system messages that they currently can't delete that we want them to delete, so i can limit it to them.

Thanks!

**Comment by dangrous** [comment]
Oh yeah, that's a good call! We should solidify which types of messages we should allow deletion of, since that will affect both the final BE and any FE solution - but then we can re review the proposals and cs can make any updates as needed.

@quinthar @miljakljajic would it be best to discuss [this](https://github.com/Expensify/App/issues/37119#issuecomment-2137803348) in Slack? Happy to start that thread!

**Comment by dangrous** [comment]
Due to refocusing, I think this is on the back burner for a bit (unless you disagree @quinthar) but we can start that discussion when we're back into it!

**Comment by dangrous** [comment]
Switching this to monthly until we have the priority to come back to it

**Comment by miljakljajic** [comment]
@dangrous how are you feeling about this one a month since the last update?

**Comment by dangrous** [comment]
Hi! Still not having the space to work on this, but I'm hoping I can look a little bit next week once the copilot project goes into implementation

**Comment by dangrous** [comment]
> One last question for the group cc @quinthar @miljakljajic - can I get a clarification on what system messages specifically we're trying to delete? I know it at least includes policy changelogs, but is there anything else?
> 
> Asking because my initial solutions allow for admins to delete all actions, and I'm wondering if we should restrict it at all. For example, they can delete IOUs. Is that fine? If not, I'd love a list of the various actions that count as system messages that they currently can't delete that we want them to delete, so i can limit it to them.
> 
> Thanks!

Okay I have a bit of time now! I wanted to bring back this question - it's certainly easiest from a technical perspective to allow admins to delete literally anything, but if we DON'T want that, I'll need to know which ones should be restricted. Let me know if you want me to bring this up in Slack somewhere too. Thanks!

**Comment by miljakljajic** [comment]
I'd raise the question in #product and we can discuss it there. Thanks @dangrous!

**Comment by dangrous** [comment]
Posted! https://expensify.slack.com/archives/C03U7DCU4/p1723657516659069

**Comment by miljakljajic** [comment]
Generally the consensus is that we shouldn't delete log lines as they make this tool less reliable for auditors/other admins and creates a potential backwards compatibility between OldDot and NewDot, as we can't delete log lines in OldDot. Closing this out.
